### DISCUSSION TITLE
✅ Examples in swagger / documentation

### DISCUSSION THREAD

**Discussion by digimuza** [RobinTail/express-zod-api#165] [Ideas]
Currently, I don't see any possible ways how I could add additional meta information to swager generated file. Do you have any suggestions?  

I really like the library so far and I would like to help :D

**Comment by RobinTail** [comment]
Hello @digimuza 

Currently when you create a `new OpenAPI()` it returns an instance that extends the `OpenApiBuilder` class from [OpenAPI library](https://github.com/metadevpro/openapi3-ts). 

```typescript
import {OpenAPI} from 'express-zod-api';

const specification = new OpenAPI({
  routing, 
  version: '1.2.3',
  title: 'Example API',
  serverUrl: 'https://example.com'
});

const yamlString = specification.getSpecAsYaml();
```

It provides you with `addExample()` method, so you can try this:

```typescript
specification.addExample(...);
```

And besides that you have a direct access to the specification properties, for example:

```typescript
specification.rootDoc
  .paths['/v1/path'].get.responses['200']
  .content['application/json'].example // and .examples
```

I did not try it yet, but there is definitely a possibility to add something to the specification.

> I really like the library so far and I would like to help :D

I'm glad to hear that, and if you have any ideas on how the necessary functionality could be implemented, please let me know.
For example, perhaps `new OpenAPI({...})` might have additional arguments to achieve it.
I hope I managed to provide you with the necessary information ^_^

**Reply by digimuza** [reply]
Yes, I know that we can do this but it's a little bit inconvenient. My suggestion would be to write a special helper that is capable to add additional meta properties to zod object itself then in swagger generation step we can look at those properties and react accordingly 

```ts
function withMeta<T extends z.ZodType<any, any, any>>(data: T, meta: {
    description?: string
    example?: T['_output']
}) {
    data._def.meta = meta
    return data
}

const myType = withMeta(z.object({ id: z.string() }), { description: "...", example: { id: "XZC" }  })
```

Example 2

```ts
factory.build({
    method: 'post',
    input: z.object({
          id: withMeta(z.string(), { example: "f7269f08-2f68-11ec-8d3d-0242ac130003" }),
          address: withMeta(z.string(), { example: "Gamyklos g. 11, LT-96155, Klaipėdos r. sav." }) 
    }),
    output: z.object({
        success: z.boolean(),
    }),
    description: "Health check endpoint",
    handler: async ({ input, options: { db, logger } }) => {
        logger.info("asd")
        return { success: true };
    }: z.object({
          id: withMeta(z.string(), { example: "f7269f08-2f68-11ec-8d3d-0242ac130003" }),
          address: withMeta(z.string(), { example: "Gamyklos g. 11, LT-96155, Klaipėdos r. sav." }) 
    }),
    output: z.object({
        success: z.boolean(),
    }),
    description: "Endpoint description",
    handler: async () => {
        return { success: true };
    }
})
```

Example 3

```ts
factory.build({
    method: 'post',
    input: withMeta(z.object({
          id: z.string(),
          address: z.string()
    }), { example: { id: "f7269f08-2f68-11ec-8d3d-0242ac130003", address: "Gamyklos g. 11, LT-96155, Klaipėdos r. sav." } }),
    output: z.object({
        success: z.boolean(),
    }),
    description: "Health check endpoint",
    handler: async ({ input, options: { db, logger } }) => {
        logger.info("asd")
        return { success: true };
    }: z.object({
          id: withMeta(z.string(), { example: "f7269f08-2f68-11ec-8d3d-0242ac130003" }),
          address: withMeta(z.string(), { example: "Gamyklos g. 11, LT-96155, Klaipėdos r. sav." }) 
    }),
    output: z.object({
        success: z.boolean(),
    }),
    description: "Endpoint description",
    handler: async () => {
        return { success: true };
    }
})
```

And then https://github.com/RobinTail/express-zod-api/blob/master/src/open-api.ts

describeSchema method we can extract type metadata

```ts
const describeSchema = (value: z.ZodTypeAny, isResponse: boolean): SchemaObject => {
  const meta = value._def.meta || {}
  // Yes we need validate it
  // And probably preprocess it
  
  const otherProps: SchemaObject = { ...meta };

  if (value.isNullable()) {
    otherProps.nullable = true;
  }
  ....
} 
```

What are your thoughts on this approach?

**Reply by RobinTail** [reply]
Wow, @digimuza , that's actually a great idea!
I like this approach. I thought about the metadata earlier, tried to extend Zod classes, but I didn't like it. 
Your solution seems to me very elegant and I will try to implement it.
Thanks a lot!

**Comment by RobinTail** [comment]
@digimuza 

I thought a little about this problem, and I realized that the only thing that confuses me is the requirement of the second parameter of `withMeta`.

I decided to try to keep the factory interface using a slightly different approach. But I'm not sure if there are any disadvantages or risks. Please take a look and tell me what you think of this solution? 

Personally, I like how the application of `withMeta` looks in the end:

```typescript
type ExampleProp<T extends z.ZodTypeAny> = T['_output'];
type DescriptionProp = string;
type WithMeta<T extends z.ZodTypeAny> = T & {
  example: (value: ExampleProp<T>) => WithMeta<T>;
  description: (description: DescriptionProp) => WithMeta<T>;
}

const withMeta = <T extends z.ZodTypeAny>(schema: T) => {
  schema._def.meta = {};
  Object.defineProperty(schema, 'example', {
    get() {
      return (example: ExampleProp<T>) => {
        this._def.meta.example = example;
        return this;
      };
    }
  });
  Object.defineProperty(schema, 'description', {
    get() {
      return (description: DescriptionProp) => {
        this._def.meta.description = description;
        return this;
      };
    }
  });
  return schema as WithMeta<T>;
};

// usage:

const myType = withMeta(z.object({
  id: z.string()
})).example({
  id: 'test'
}).description('something');
```

**Reply by digimuza** [reply]
@RobinTail  your solution looks sick 👍  I don't see any immediate problems. One thing that I thought that 

 schema._def.meta is super generic variable name it would be wise to change it by using 
 schema._def['express-zod-api'] in this way package will have unique namespace 
 
 Zod ecosystem is evolving rapidly and question what will happen if zod package maintainer decides to add meta prop  in schema._def?

**Reply by RobinTail** [reply]
make sense. Thank you @digimuza

**Comment by RobinTail** [comment]
While I've been working on the implementation the ability to add description to the schema has been released in Zod 3.11.x
I'm surprised because the author multiple times explained that he is against this feature.

**Reply by digimuza** [reply]
Yes I also little bit confused because he said multiple times that he will not implement such feature :D

**Reply by digimuza** [reply]
@RobinTail do you need any help?

**Reply by RobinTail** [reply]
Thank you @digimuza , 
Looks like I can handle it. This turned out to be a little more complex in general case, as it requires merging the middleware and endpoint examples. Overall, everything is going well. I prepared a pull request #175, which I am still working on. If you have any thoughts on this, feel free to add comments there as well.

**Comment by digimuza** [comment]
@RobinTail  You are amazing ❤️ ❤️ ❤️ ❤️ ❤️

**Comment by digimuza** [comment]
One additional thing that would be amazing is ability to generate models that can be referenced across multiple endpoints. This is necessary if you want generate any useful API client with open-api generators. 

Is it possible to have this kinda feature?

**Reply by RobinTail** [reply]
Could you please clarify: what are those models? Is there any example or sample invocation of such a functionality? @digimuza

**Comment by digimuza** [comment]
For example, we have this swagger documentation 
https://petstore.swagger.io/

If we try to generate typescript-axios client with https://www.npmjs.com/package/@openapitools/openapi-generator-cli

We get a nice API client

Try to generate typescript-axios  client from a schema that is generated by express-zod-api and you will see problem

**Comment by RobinTail** [comment]
@digimuza 
I'd like to inform you that the new feature — Client generation — has been released in version 6.1.0
